// @tinkoff/request-plugin-cache-memory/lib/memory.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var prop = require('@tinkoff/utils/object/prop');
var propOr = require('@tinkoff/utils/object/propOr');
var requestCore = require('@tinkoff/request-core');
var requestCacheUtils = require('@tinkoff/request-cache-utils');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var prop__default = /*#__PURE__*/_interopDefaultLegacy(prop);
var propOr__default = /*#__PURE__*/_interopDefaultLegacy(propOr);

/**
 * determine how long should we wait for background requests relative to the timeout of the initial request
 */
const BACKGROUND_REQUEST_TIMEOUT_FACTOR = 2;
const getStaleBackgroundRequestTimeout = (params) => {
    const { requestTimeout, configTimeout } = params;
    if (configTimeout) {
        return configTimeout;
    }
    if (requestTimeout) {
        return requestTimeout * BACKGROUND_REQUEST_TIMEOUT_FACTOR;
    }
    return undefined;
};

const CACHE_DEFAULT_TTL = 300000; // 5 min
/**
 * Caches requests response into memory.
 * Uses library `@tinkoff/lru-cache-nano` as memory storage.
 *
 * requestParams:
 *      memoryCache {boolean} - disable this plugin at all
 *      memoryCacheForce {boolean} - plugin will only be executed on complete phase
 *      memoryCacheTtl {number} - ttl of cache of the current request
 *      memoryCacheAllowStale {boolean} - is allowed to use outdated value from cache
 *          (if true outdated value will be returned and request to update it will be run in background)
 *
 * metaInfo:
 *      memoryCache {boolean} - is current request was returned from this cache
 *      memoryCacheOutdated {boolean} - is value in cache is outdated (only for plugin with allowStale = true)
 *
 * @param {object} [lruOptions = {max: 1000, ttl: 300000}] - options passed to @tinkoff/lru-cache-nano library
 * @param {boolean} [shouldExecute = true] is plugin activated by default
 * @param {boolean} [allowStale = false] is allowed to use outdated value from cache
 *      (if true outdated value will be returned and request to update it will be run in background)
 * @param {number} [staleTtl = lruOptions.ttl] time in ms while outdated value is preserved in cache while
 *       executing background update
 * @param {number} [staleBackgroundRequestTimeout] time in ms for background request timeout for the stale response
 * @param {function} memoryConstructor cache factory
 * @param {function} getCacheKey function used for generate cache key
 */
var memory = ({ lruOptions = { max: 1000, ttl: CACHE_DEFAULT_TTL }, shouldExecute = true, allowStale = false, staleTtl = 'ttl' in lruOptions ? lruOptions.ttl : CACHE_DEFAULT_TTL, staleBackgroundRequestTimeout = undefined, memoryConstructor = (options) => new (require('@tinkoff/lru-cache-nano'))(options), getCacheKey = undefined, } = {}) => {
    const lruCache = memoryConstructor({
        ...lruOptions,
        allowStale: true, // should be true for the opportunity to control it for individual requests
    });
    return {
        shouldExecute: requestCacheUtils.shouldCacheExecute('memory', shouldExecute),
        init: (context, next, makeRequest) => {
            const cacheKey = requestCacheUtils.getCacheKey(context, getCacheKey);
            if (lruCache.has(cacheKey)) {
                context.updateExternalMeta(requestCacheUtils.metaTypes.CACHE, {
                    memoryCache: true,
                    memoryCacheOutdated: false,
                });
                return next({
                    status: requestCore.Status.COMPLETE,
                    response: lruCache.get(cacheKey),
                });
            }
            const allowOutdated = propOr__default["default"]('memoryCacheAllowStale', allowStale, context.getRequest());
            const outdated = allowOutdated && lruCache.peek(cacheKey);
            if (outdated) {
                const request = context.getRequest();
                context.updateExternalMeta(requestCacheUtils.metaTypes.CACHE, {
                    memoryCache: true,
                    memoryCacheOutdated: true,
                });
                lruCache.set(cacheKey, outdated, { ttl: staleTtl }); // remember outdated value, to prevent losing it
                setTimeout(() => makeRequest({
                    ...request,
                    timeout: getStaleBackgroundRequestTimeout({
                        requestTimeout: request.timeout,
                        configTimeout: staleBackgroundRequestTimeout,
                    }),
                    memoryCacheForce: true,
                    memoryCacheBackground: true,
                    // ignore any abort properties to the request in the background
                    // as if abortSignal aborts background request we won't update cache
                    // and will try to update it in next requests over and over again.
                    // Also we are not blocking main execution anyway
                    // to care about execution timings in that case
                    signal: undefined,
                    abortPromise: undefined
                }).catch(() => {
                    // do nothing here because sub request have all logging stuff
                }), 15); // run background request to update cache
                return next({
                    status: requestCore.Status.COMPLETE,
                    response: outdated,
                });
            }
            next();
        },
        complete: (context, next) => {
            const cacheKey = requestCacheUtils.getCacheKey(context, getCacheKey);
            const ttl = prop__default["default"]('memoryCacheTtl', context.getRequest());
            lruCache.set(cacheKey, context.getResponse(), { ttl });
            context.updateExternalMeta(requestCacheUtils.metaTypes.CACHE, {
                memoryCacheBackground: prop__default["default"]('memoryCacheBackground', context.getRequest()),
            });
            next();
        },
    };
};

exports["default"] = memory;
