// @tinkoff/request-plugin-cache-deduplicate/lib/deduplicate.es.js
import { shouldCacheExecute, getCacheKey, metaTypes } from '@tinkoff/request-cache-utils';

/**
 * Deduplicate requests with equal cache keys before making a request
 *
 * requestParams:
 *      deduplicateCache {boolean} - disable this plugin at all
 *      deduplicateCacheForce {boolean} - plugin will only be executed on complete phase
 *
 * metaInfo:
 *      deduplicated {boolean} - is current request was deduplicated (is not set for the first request of equals requests)
 *
 * @param {boolean} [shouldExecute = true] is plugin activated by default
 * @param {function} getCacheKey function used for generate cache key
 */
var deduplicate = ({ shouldExecute = true, getCacheKey: getCacheKey$1 = undefined } = {}) => {
    const activeRequests = {};
    const traverseActiveRequests = (context) => {
        const state = context.getState();
        const deduplicationKey = getCacheKey(context, getCacheKey$1);
        if (deduplicationKey && activeRequests[deduplicationKey]) {
            const arr = activeRequests[deduplicationKey];
            delete activeRequests[deduplicationKey];
            arr.forEach((next) => {
                next({
                    status: state.status,
                    response: state.response,
                    error: state.error,
                });
            });
        }
    };
    return {
        shouldExecute: shouldCacheExecute('deduplicate', shouldExecute),
        init: (context, next) => {
            const deduplicationKey = getCacheKey(context, getCacheKey$1);
            if (activeRequests[deduplicationKey]) {
                context.updateExternalMeta(metaTypes.CACHE, {
                    deduplicated: true,
                });
                activeRequests[deduplicationKey].push(next);
                return;
            }
            activeRequests[deduplicationKey] = [];
            next();
        },
        complete: (context, next) => {
            traverseActiveRequests(context);
            next();
        },
        error: (context, next) => {
            traverseActiveRequests(context);
            next();
        },
    };
};

export { deduplicate as default };
